File: ice0208/linked-list/328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head):
        if head == None or head.next == None:
            return head

        odd_head = head
        odd_tail = head
        even_head = head.next
        even_tail = head.next

        cursor = even_tail
        isOdd = True
        while cursor.next:
            cursor = cursor.next
            if isOdd:
                odd_tail.next = cursor
                odd_tail = cursor
            else:
                even_tail.next = cursor
                even_tail = cursor

            isOdd = not isOdd

        odd_tail.next = even_head
        even_tail.next = None

        return odd_head

File: moonpoison/linked_list/328.cpp
#include<bits/stdc++.h>
using namespace std;
struct ListNode 
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) 
    {
        if(!head) return NULL;
        ListNode *odd=head, *even=head->next, *first=even;
        while(even!=nullptr&&even->next!=nullptr)
        {
            odd->next=even->next;
            odd=odd->next;
            even->next=odd->next;
            even=even->next;
        }
        odd->next=first;
        return head;
    }
};